### PR TITLE
Add tests for invalid unbraced unicode escapes and for valid unassigned unicode escapes

### PR DESCRIPTION
**Pull request by SiriusStarr**
[Per conversation on Discourse.](https://discourse.dhall-lang.org/t/dhall-haskell-not-conforming-to-the-standard-with-respect-to-disallowed-unicode-characters/90/5)  Needs looking over to confirm this is what people had in mind.

### CODE REVIEW

File: tests/parser/success/nonAssignedUnicodeB.diag [hunk 1]
```diff
@@ -0,0 +1 @@
+[18, "🿰"]
```
Comment (sjakobi) [inline tests/parser/success/nonAssignedUnicodeB.diag]: According to Hydra, this is the actual output of `cbor2diag` for the `dhallb`:

```suggestion
[18, "\u{1FFF0}"]
```

Comment (SiriusStarr) [reply tests/parser/success/nonAssignedUnicodeB.diag]: So if you run `dhall encode --json` with a "valid" unicode escape, it produces this:

```
dhall encode --json <<< '"\u5564"'
[
    18,
    "啤"
]
```
which was how you'd suggested I generate it.  But yeah it looks like the ones in the actual test cases are in there as unicode escapes, so I changed it.

**Comment by SiriusStarr** [comment]
@sjakobi the relevant portion of the grammar states:
```
double-quote-escaped =
      %x22                 ; '"'    quotation mark  U+0022
    / %x24                 ; '$'    dollar sign     U+0024
    / %x5C                 ; '\'    reverse solidus U+005C
    / %x2F                 ; '/'    solidus         U+002F
    / %x62                 ; 'b'    backspace       U+0008
    / %x66                 ; 'f'    form feed       U+000C
    / %x6E                 ; 'n'    line feed       U+000A
    / %x72                 ; 'r'    carriage return U+000D
    / %x74                 ; 't'    tab             U+0009
    / %x75 unicode-escape  ; 'uXXXX' / 'u{XXXX}'    U+XXXX

; The parser must also reject Unicode escape sequences that are either:
;
; * Surrogate pairs (i.e. `%xD800-DFFF`)
; * Non-characters (i.e. `%xNFFFE-xNFFFF` for each `N` in `{ 0 .. F }`)
;
; See the `valid-non-ascii` rule for the exact ranges that are not allowed
unicode-escape = 4HEXDIG / "{" 1*HEXDIG "}"
```
so it does not need updating, at least as I read it, since it says that valid unicode escapes are `\uHHHH` or `\u{H...}` where `HHHH`/`H...` is not in the ranges `0xD800-DFFF` or `0xNFFFE-NFFFF` for each `N` in `{ 0 .. F }`.

**Comment by SiriusStarr** [comment]
I went ahead and moved the success test into the `text` subfolder, since it seems to belong there.

**Comment by sjakobi** [comment]
> ```
> ; The parser must also reject Unicode escape sequences that are either:
> ;
> ; * Surrogate pairs (i.e. `%xD800-DFFF`)
> ; * Non-characters (i.e. `%xNFFFE-xNFFFF` for each `N` in `{ 0 .. F }`)
> ;
> ; See the `valid-non-ascii` rule for the exact ranges that are not allowed
> unicode-escape = 4HEXDIG / "{" 1*HEXDIG "}"
> ```

I guess we could try to move these exceptions from the comments into the rule, but I'm not sure whether it's worth the effort. Having tests is more valuable IMHO!

Cheers! :)

**Comment by SiriusStarr** [comment]
@sjakobi I put together a first shot at a grammar change to actually make this explicit.  This changes what's valid in the following ways:

* Codepoints past Plane 16 are no longer valid (e.g. `"\u{DDDDDDDDDDDDDDDDDDDDDDDD}"` is no longer valid Dhall).  This shouldn't be a problem since those were never valid characters.
* Plane 16 non-characters (`0x10FFFE-10FFFD`) are now correctly excluded, since the comment only specifically excluded non-characters in Planes 0-15.

Ridiculous amounts of zero-padding are still allowed, e.g. `"\u{000000000000000000000000000000030}"` is still valid Dhall.

File: standard/dhall.abnf [hunk 2]
```diff
@@ -257,9 +257,38 @@ double-quote-escaped =
...
+;
+; * Surrogate pairs (i.e. `%xD800-DFFF`)
+; * Non-characters (i.e. `%xNFFFE-NFFFF` for each `N` in `{ 0 .. F }` or `%x10FFFE-10FFFF`)
+;
+valid-braced-codepoint = 1*3HEXDIG ; %x000-FFF
+                       / valid-unbraced-escape ; %x0000-FFFD (excluding surrogate pairs) (Plane 0)
+                       / ("1" / "2" / "3" / "4" / "5" / "6" / "7" / "8" / "9" / "A" / "B" / "C" / "D" / "E" / "F" / "10") valid-unicode-suffix; %xN0000-NFFFD for each `N` in `{ 0 .. F }` (Planes 1-15) and %x100000-10FFFD (Plane 16)
+
+; Allow zero padding valid codepoints
+valid-braced-escape = *"0" valid-braced-codepoint
```
Comment (sjakobi) [inline standard/dhall.abnf]: Hmm, I guess this is more explicit and more digestible by parser generators, but I personally find it harder to understand.

I guess we don't have a construct to "subtract" a set of invalid tokens? Could we add one? I'm thinking of something like
`Alex`'s [set-difference operator for character sets](https://www.haskell.org/alex/doc/html/charsets.html):

> `set0 # set1`
>
>    Matches all the characters in `set0` that are not in `set1`.

Comment (SiriusStarr) [reply standard/dhall.abnf]: To the best of my knowledge, ABNF does not allow for "subtractive" rules in any way.  We obviously could make up one, but then it wouldn't be ABNF, which defeats the point of specifying the grammar in ABNF.

I've significantly improved the comments around the rule clarifying what exactly it means; take a look and tell me if you think it can be improved further.  I personally think that trying to introduce our own subtractive specification that doesn't conform to an extant standard would probably be *more* confusing to someone reading it than just a really good comment.

Comment (SiriusStarr) [reply standard/dhall.abnf]: Dropped the prefixes and got rid of the `hex-digit-through-d` rule since it was just confusing.  Also reformatted some multi-line things to more closely match other places in the grammar.

I had just included the `valid-` prefixes by analogy the the `valid-non-ascii` rule.
